File: server/src/engine/systems/movementsystem.ts
import EntityManager = require("../entitymanager");
import WorldSystem = require("./worldsystem");
import Vector2 = require("../geometry/vector2");
import Transform = require("../components/transform");

class MovementSystem implements System{
    static maxTileVolume = 100;
    
    constructor(private entityManager: EntityManager, private world: WorldSystem) {
    }

    newCla(a: number): Cla {
        return {a: a};
    }

    moveEntity(entity: number, destination: Vector2): boolean {
        var position = this.entityManager.getComponent(entity, Transform).position;
        var startingTile = this.world.tiles[position.x][position.y];
        var finishTile = this.world.tiles[destination.x][destination.y];
        if (finishTile.volume + this.entityManager.getComponent(entity, Transform).volume < MovementSystem.maxTileVolume) {
            position.x = destination.x;
            position.y = destination.y;
            this.world.moveEntity(entity, position);
            return true;
        }
        return false;
    }

}

export = MovementSystem;

File: sharpServer/Engine/Systems/MovementSystem.cs
namespace SharpServer.Engine.Systems
{
    internal class MovementSystem
    {
        private EntityManager entityManager;
        private WorldSystem worldSystem;

        public MovementSystem(EntityManager entityManager, WorldSystem worldSystem)
        {
            this.entityManager = entityManager;
            this.worldSystem = worldSystem;
        }

        public bool moveEntity(int entity, Vector2 destination)
        {
            var position = entityManager.getComponent<Transform>(entity).position;
            var startingTile = worldSystem.tiles[position.x, position.y];
            var finishTile = worldSystem.tiles[destination.x, destination.y];
            if (finishTile.volume + entityManager.getComponent<Shape>(entity).volume < Tile.maxVolume)
            {
                position.x = destination.x;
                position.y = destination.y;
                worldSystem.moveEntity(entity, position);
                return true;
            }
            return false;
        }
    }
}